flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSourceSinkFactory.java

```diff
@@ -0,0 +1,320 @@
+		.key("scan.partition.upper-bound")
+		.longType()
+		.noDefaultValue()
+		.withDescription("the largest value of the last partition.");
+	private static final ConfigOption<Integer> SCAN_FETCH_SIZE = ConfigOptions
+		.key("scan.fetch-size")
+		.intType()
+		.defaultValue(0)
+		.withDescription("gives the reader a hint as to the number of rows that should be fetched, from" +
+			" the database when reading per round trip. If the value specified is zero, then the hint is ignored. The" +
+			" default value is zero.");
+
+	// look up config options
+	private static final ConfigOption<Long> LOOKUP_CACHE_MAX_ROWS = ConfigOptions
+		.key("lookup.cache.max-rows")
+		.longType()
+		.defaultValue(-1L)
+		.withDescription("the max number of rows of lookup cache, over this value, the oldest rows will " +
+			"be eliminated. \"cache.max-rows\" and \"cache.ttl\" options must all be specified if any of them is " +
+			"specified. Cache is not enabled as default.");
+	private static final ConfigOption<Duration> LOOKUP_CACHE_TTL = ConfigOptions
+		.key("lookup.cache.ttl")
+		.durationType()
+		.defaultValue(Duration.ofSeconds(10))
+		.withDescription("the cache time to live.");
+	private static final ConfigOption<Integer> LOOKUP_MAX_RETRIES = ConfigOptions
+		.key("lookup.max-retries")
+		.intType()
+		.defaultValue(3)
+		.withDescription("the max retry times if lookup database failed.");
+
+	// write config options
+	private static final ConfigOption<Integer> SINK_FLUSH_MAX_ROWS = ConfigOptions
+		.key("sink.flush.max-rows")
+		.intType()
+		.defaultValue(5000)
+		.withDescription("the flush max size (includes all append, upsert and delete records), over this number" +
+			" of records, will flush data. The default value is 5000.");
+	private static final ConfigOption<Long> SINK_FLUSH_INTERVAL = ConfigOptions
+		.key("sink.flush.interval")
+		.longType()
+		.defaultValue(0L)
+		.withDescription("the flush interval mills, over this time, asynchronous threads will flush data. The " +
+			"default value is 0, which means no asynchronous flush thread will be scheduled.");
+	private static final ConfigOption<Integer> SINK_MAX_RETRIES = ConfigOptions
+		.key("sink.max-retries")
+		.intType()
+		.defaultValue(3)
+		.withDescription("the max retry times if writing records to database failed.");
+
+	@Override
+	public DynamicTableSink createDynamicTableSink(Context context) {
+		final FactoryUtil.TableFactoryHelper helper = FactoryUtil.createTableFactoryHelper(this, context);
+		final ReadableConfig config = helper.getOptions();
+
+		helper.validate();
+		validateConfigOptions(config);
+		JdbcOptions jdbcOptions = getJdbcOptions(config);
+		TableSchema physicalSchema = TableSchemaUtils.getPhysicalSchema(context.getCatalogTable().getSchema());
+
+		return new JdbcDynamicTableSink(
+			jdbcOptions,
+			getJdbcExecutionOptions(config),
+			getJdbcDmlOptions(jdbcOptions, context.getCatalogTable().getSchema()),
```

Comment:
Should use `physicalSchema` here. 
We should keep primary key in `TableSchemaUtils.getPhysicalSchema`. ES connector also did this: https://github.com/apache/flink/pull/12184/files#diff-8fe9a415913cbd47651e39f66af0114e